Double-stranded DNA is able to form triple-helical structures by accommodating a third nucleotide strand in its major groove. This sequence-specific process offers a potent mechanism for targeting genomic loci of interest, that is of great interest for bio-technological and gene-therapeutic applications.

Triplex-formation requires a duplex, presumably DNA, and a single-stranded nucleotide sequence acting as the third strand, presumably RNA.

The process of predicting putative triplexes in sequence data can be divided into three steps:

  1. identifying the triplex-forming oligonucleotides (TFO) in single-stranded sequences
  1. identifying the triplex target sites in double-stranded sequences able to accomondate a third strand
  1. assessing the compatibility of potential TFO/TTS pairs according to the canonical triplex formation rules

Triplexator, short for triple-helix locator, employs approximate pattern matching based on the triplex formation rules in order to find putative TFO-TTS pairs (triplexes) in two sets of nucleotide sequences, where one set contains the duplex sequences and the other the single-stranded partner.

More information about Triplexator are available from http://bioinformatics.org.au/triplexator